### PR TITLE
Input event improvements

### PR DESCRIPTION
**Pull request by Swiggies**
This PR will allow developers to use a new InputEventState class to subscribe to input events which pass in the state of the input, where they can then handle what happens per-state, rather than having to make multiple input actions, and then creating multiple InputEvents.

This is pretty similar to my last input PR but just adds the functionality to InputEvent (through a new class).

Decided on making it a new C# class as I didn't want to break current functionality with InputEvents but open to suggestions.

Example usage:
```csharp
public InputEventState inputEvent = new InputEventState("Jump");

public override void OnEnable()
{
    // Here you can add code that needs to be called when script is enabled (eg. register for events)
    inputEvent.Triggered += OnTriggeredState;
}

private void OnTriggeredState(InputActionState state)
{
    if (state == InputActionState.Press) Debug.Log("Press");
    if (state == InputActionState.Release) Debug.Log("Release");
}

public override void OnDisable()
{
    // Here you can add code that needs to be called when script is disabled (eg. unregister from events)
    inputEvent.Triggered -= OnTriggeredState;
}
```

### CODE REVIEW

**Comment by Tryibion** [comment]
Just a thought for discussion. Would it make sense to have a different event/Action per state?

File: Source/Engine/Engine/InputEventState.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        /// <summary>
+        /// Returns true if the event has been triggered during the current frame (e.g. user pressed a key). Use <see cref="Triggered"/> to catch events without active waiting.
+        /// </summary>
+        public bool Active => Input.GetAction(Name);
+
+        /// <summary>
+        /// Occurs when event is triggered (e.g. user pressed a key). Called before scripts update.
+        /// Has an additional parameter to read the input state.
+        /// </summary>
+        public event Action<InputActionState> Triggered;
```
Comment (GoaLitiuM) [inline Source/Engine/Engine/InputEventState.cs:27]: So why can't this be under `InputEvent` as `public event Action<InputActionState> StateChanged;` or something like that? I don't think adding a new action would be a breaking change.

Comment (Swiggies) [reply Source/Engine/Engine/InputEventState.cs:27]: I can have a look at adding this, but normal InputEvents are handled the same way as regular InputActions and are only activated when in an exact state, the one that's specified in Input Settings. I initially tried to do this but I can have another look at adding it that way.

**Comment by Swiggies** [comment]
> Just a thought for discussion. Would it make sense to have a different event/Action per state?

This is how Unity handles it, somewhat. I'm happy to implement that if it makes more sense to others.

**Comment by mafiesto4** [comment]
@Swiggies very cool

Do you think it's better to use this version or #1496 ? I'm fine with this small API breakage.

**Comment by Tryibion** [comment]
> @Swiggies very cool
> 
> Do you think it's better to use this version or #1496 ? I'm fine with this small API breakage.

I assume you mean #1529

**Comment by mafiesto4** [comment]
Ah yeah, sorry my bad #1529

**Comment by mafiesto4** [comment]
Thanks again @Swiggies - I merged code from @Tryibion to keep the API slicker (less types) - but I'll include you as a contributor in the changelog.

**Pull request closed without merging**